test/Microsoft.Sbom.Parsers.Spdx30SbomParser.Tests/Utils/SPDXExtensionsTest.cs

```diff
@@ -0,0 +1,208 @@
+using Microsoft.Sbom.Parsers.Spdx30SbomParser.Utils;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Sbom.Utils;
+
+[TestClass]
+public class SPDXExtensionsTest
+{
+    private const string PackageSpdxIdWithMissingPackageId = "SPDXRef-Package-D99F4826E74FCF3A40C17AF7F82ED84E673D7D70975A6337C30F81C71E0C2BAC";
+    private const string PackageSpdxIdWithPackageId = "SPDXRef-Package-D59091CB205934C58034F3FB2290A10F3DC6AFDA895B099066F9D40245A239C2";
+    private const string PackageSpdxIdWithVersionAndMissingPackageId = "SPDXRef-Package-76A4881BB39276EA67CA840875A0EDFFE7562057AD19998491518B222D278E87";
+    private const string ExternalReferenceSpdxId = "DocumentRef-externalRefName-sha1Value";
+    private const string FileSpdxId = "SPDXRef-File-filePath-sha1Value";
+    private const string ElementSpdxId = "SPDXRef-Element-82A3537FF0DBCE7EEC35D69EDC3A189EE6F17D82F353A553F9AA96CB0BE3CE89";
+
+    private InternalSbomFileInfo fileInfo = null;
+    private SbomPackage packageInfo = null;
+
+    private File spdxFile = null;
+    private Package spdxPackage = null;
+    private ExternalMap externalMap = null;
+    private IEnumerable<Checksum> checksums = new List<Checksum>
+    {
+        new Checksum
+        {
+            Algorithm = AlgorithmName.SHA1,
+            ChecksumValue = "sha1Value",
+        },
+        new Checksum
+        {
+            Algorithm = AlgorithmName.SHA256,
+            ChecksumValue = "sha256Value",
+        }
+    };
+
+    private IEnumerable<Checksum> checksumsWithoutSHA1 = new List<Checksum>
+    {
+        new Checksum
+        {
+            Algorithm = AlgorithmName.SHA256,
+            ChecksumValue = "sha256Value",
+        }
+    };
+
+    [TestInitialize]
+    public void Setup()
+    {
+        spdxPackage = new Package
+        {
+            Name = "packageName",
+            PackageVersion = "1.0.0"
+        };
+
+        packageInfo = new SbomPackage
+        {
+            Id = "packageId",
+            Type = "npm",
+            PackageName = "packageName",
+            PackageVersion = "1.0.0"
+        };
+
+        externalMap = new ExternalMap();
+
+        fileInfo = new InternalSbomFileInfo()
+        {
+            Path = "filePath",
+            Checksum = checksums
+        };
+
+        spdxFile = new File();
+    }
+
+    [TestMethod]
+    public void AddSpdxIdToPackage_PackageInfoIsNull_Throws()
+    {
+        Assert.ThrowsException<ArgumentNullException>(() => spdxPackage.AddSpdxId(null));
+    }
+
+    [DataRow(true)]
+    [DataRow(false)]
+    [TestMethod]
+    public void AddSpdxIdToPackage_PackageInfoWithId_Success(bool usePackageWithVersion)
+    {
+        if (usePackageWithVersion)
+        {
+            spdxPackage.AddSpdxId(packageInfo);
+        }
+        else
+        {
+            packageInfo.PackageVersion = null;
+            spdxPackage.AddSpdxId(packageInfo);
+        }
+
+        Assert.AreEqual(PackageSpdxIdWithPackageId, spdxPackage.SpdxId);
```

Comment:
Wait... shouldn't the SpdxId change if PackageVersion changes?

Reply:
No, per my understanding - If the package id is provided, then the package version is not used when creating the SPDX id. So regardless of the version, the SPDX id should be the same if the package id is the same. I also cross checked that SPDX 2.2 has the same behavior.